cmake/KokkosConfigCommon.cmake.in

```diff
@@ -209,6 +209,290 @@ int main()
+    if(compile_dependency)
+      list(APPEND compile_targets ${target})
+    endif()
+    if(link_dependency)
+      list(APPEND link_targets ${target})
+    endif()
+  endforeach()
+
+  # push results out of recursive function
+  set(${FOUND_COMPILE_TARGETS} ${compile_targets} PARENT_SCOPE)
+  set(${FOUND_LINK_TARGETS} ${link_targets} PARENT_SCOPE)
+endfunction()
+
+# this function sets the source and target properties for all targets that link to a given library
+#
+#       PUBLIC          --> input: set downstream sources and targets to have a public dependency on kokkos. If not given a private dependency is assumed
+#       LIBRARY         --> input: name of the library downstream targets need to link to for this function to set the properties.
+#       USE_LAUNCHER    --> input: Command for a launcher to use
+#
+function(kokkos_set_dependent_library_properties)
+  cmake_parse_arguments(INP "PUBLIC" "LIBRARY" "USE_LAUNCHER" ${ARGN})
+
+  if(NOT INP_LIBRARY)
+    message(FATAL_ERROR "'kokkos_set_dependent_library_properties' needs a LIBRARY argument")
+  endif()
+
+  kokkos_get_directory_targets(COMPILE_TARGETS LINK_TARGETS ${INP_LIBRARY} ${PROJECT_SOURCE_DIR})
+
+  if(INP_PUBLIC)
+    if(@Kokkos_ENABLE_COMPILE_AS_CMAKE_LANGUAGE@)
+      foreach(target IN LISTS COMPILE_TARGETS)
+        get_target_property(target_sources ${target} SOURCES)
+        get_target_property(target_dir ${target} SOURCE_DIR)
+        list(FILTER target_sources INCLUDE REGEX "\\.cpp$|\\.cxx$")
+        message(
+          DEBUG
+          "Kokkos setting language to @KOKKOS_COMPILE_LANGUAGE@ on target ${target} for sources ${target_dir}/{${target_sources}}"
+        )
+        foreach(source IN LISTS target_sources)
+          set_property(
+            SOURCE ${target_dir}/${source} TARGET_DIRECTORY ${target} PROPERTY LANGUAGE @KOKKOS_COMPILE_LANGUAGE@
+          )
+        endforeach()
+      endforeach()
+    endif()
+    kokkos_set_source_and_target_properties(USE_LAUNCHER ${INP_USE_LAUNCHER} TARGET ${LINK_TARGETS})
+  endif()
+
+  kokkos_set_source_and_target_properties(USE_LAUNCHER ${INP_USE_LAUNCHER} LINK_ONLY_TARGET ${LINK_TARGETS})
+endfunction()
+
+# this function sets the source and target properties for all targets that link to a given library but deferred to the end of the cmake runtime
+#
+#       PUBLIC          --> input: set downstream sources and targets to have a public dependency on kokkos. If not given a private dependency is assumed
+#       LIBRARY         --> input: name of the library downstream targets need to link to for this function to set the properties.
+#       USE_LAUNCHER    --> input: Command for a launcher to use
+#
+function(kokkos_defer_set_dependent_library_properties)
+  cmake_parse_arguments(ARG "PUBLIC" "LIBRARY" "USE_LAUNCHER" ${ARGN})
+
+  if(NOT ARG_LIBRARY)
+    message(FATAL_ERROR "'kokkos_defer_set_dependent_library_properties' needs a LIBRARY argument")
+  endif()
```

Comment:
this is repeated here prevent running into this [issue with cmake](https://gitlab.kitware.com/cmake/cmake/-/issues/26599)